src/net/socket.rs

```diff
@@ -26,51 +31,72 @@ impl SocketSystem {
    }
 
    /// Resolves the socket addresses the given hostname could refer to.
-   async fn resolve_address_with_default_port(hostname: &str) -> anyhow::Result<Vec<SocketAddr>> {
-      if let Ok(addresses) = lookup_host(hostname).await {
-         Ok(addresses.collect())
-      } else {
-         Ok(lookup_host((hostname, relay::DEFAULT_PORT)).await?.collect())
+   async fn resolve_address_with_default_port(hostname: &str) -> anyhow::Result<url::Url> {
+      let mut url = url::Url::parse(&format!("ws://{}", hostname))?;
+
+      if let None = url.port() {
+         // Url::set_port on Error does nothing, so it is fine to ignore it
+         #[allow(unused_must_use)]
+         {
+            url.set_port(Some(relay::DEFAULT_PORT));
+         }
       }
+
+      Ok(url)
    }
 
    async fn connect_inner(self: Arc<Self>, hostname: String) -> anyhow::Result<Socket> {
-      let addresses = Self::resolve_address_with_default_port(&hostname)
+      let address = Self::resolve_address_with_default_port(&hostname)
          .await
-         .context("Could not resolve address. Are you sure the IP is correct?")?;
-      log::info!("resolved addresses: {:?}", addresses);
-      let stream = TcpStream::connect(addresses.as_slice()).await?;
-      stream.set_nodelay(true)?;
-      let (mut read_half, write_half) = stream.into_split();
+         .context("Could not resolve address. Are you sure the address is correct?")?;
+      let (stream, _) = connect_async(address).await?;
+      let (sink, mut stream) = stream.split();
       log::info!("connection established");
 
-      let version = read_half.read_u32().await?;
+      let version = if let Some(version) = stream.next().await {
+         match version {
+            Ok(Message::Binary(version)) => {
+               let array: [u8; 4] = match version.try_into() {
+                  Ok(a) => a,
+                  Err(_) => anyhow::bail!("Relay sent invalid version."),
+               };
+
+               u32::from_le_bytes(array)
+            }
+            Ok(_) => anyhow::bail!("Relay sent invalid packet."),
+            Err(e) => anyhow::bail!(e),
+         }
+      } else {
+         anyhow::bail!("Didn't received version packet.")
```

Comment:
In English the infinitive form is used after `did` or `didn't`.
```suggestion
         anyhow::bail!("Didn't receive version packet.")
```